### PR TITLE
Initial Windows support

### PR DESCRIPTION
**Pull request by albinekb**
#37 without linux support

### todo
- [ ] Audio/audiosources
   http://zornsoftware.codenature.info/blog/record-screen-audio-with-ffmpeg.html
- [ ] Split startRecording functions per OS
- [x] [Figure out how to get ffmpeg (Should we ship the ffmpeg binary (.exe)?)](https://github.com/wulkano/aperture/pull/37#issuecomment-307482656)
  >No, that's too heavy and outside the scope of this plugin. We can add a ffmpegBinary option instead.

### CODE REVIEW

**Comment by albinekb** [comment]
@sindresorhus `.mp4` was working, but windows media player can't play it 😄  tested with vlc and it works.

File: README.md [hunk 1]
```diff
@@ -11,13 +11,16 @@
 $ npm install aperture
 ```
 
-*Requires macOS 10.10 or later.*
+requires: *macOS: 10.10 or later.* | *Windows: ffmpeg*
```
Comment (sindresorhus) [inline README.md]: Should be:

```
*Requires minimum macOS 10.10 or Windows 7.*
```

File: README.md [hunk 2]
```diff
@@ -11,13 +11,16 @@
...
 
 
 ## Usage
 
 ```js
-const aperture = require('aperture')();
+const aperture = require('aperture')({
+  // ffmpegBinary is *required* on Windows
```
Comment (sindresorhus) [inline README.md]: This comment is moot. It's already explained in the docs section.

File: README.md [hunk 3]
```diff
@@ -42,7 +45,16 @@ See [`example.js`](example.js) if you want to quickly try it out. *(The example
...
 
-### instance = aperture()
+### instance = aperture([options])
+
+#### options
+
+##### ffmpegBinary
+
+Type: `string`<br>
+**Required on windows**
```
Comment (sindresorhus) [inline README.md]: windows => Windows

File: index.js [hunk 4]
```diff
@@ -50,21 +78,43 @@ class Aperture {
...
+
+        this.recorder = execa(path.join(__dirname, 'swift', 'main'), recorderOpts);
+      } else if (IS_WINDOWS) {
+        const ffmpegArgs = [];
+
+        if (typeof cropArea === 'object') {
+          ffmpegArgs.push(
+            '-video_size', `${cropArea.width}x${cropArea.height}`,
+            '-f', 'gdigrab',
+            '-i', 'desktop',
```
Comment (sindresorhus) [inline index.js]: The above two arguments can be in the initial array as they appear the same in both conditions.

File: README.md [hunk 5]
```diff
@@ -1,6 +1,6 @@
 <p>
   <h1 align="center">Aperture</h1>
-  <h4 align="center">Record the screen on macOS</h4>
+  <h4 align="center">Record the screen on macOS & Windows</h4>
```
Comment (sindresorhus) [inline README.md]: You need to update the package.json description too.

**Comment by albinekb** [comment]
Since `getAudioSources` is outside the class now, how do you want to handle the `ffmpegBinary` there?

**Comment by albinekb** [comment]
Is there any interest in finishing this? I don't use Windows and never will use this lib on windows, but if there are people that want this I could pick it up and finish it. Thoughts?

**Comment by sindresorhus** [comment]
Not really. I only care about the macOS part. I just reviewed this to be nice.

**Comment by CodeTheory** [comment]
For now, I don't think we can maintain this without a contributor that has windows. I'll be building a PC at some point this year and then I can maintain it, but for now I don't think we can support it properly. 😢

**Comment by albinekb** [comment]
I just did this to be nice too 😄  But as I use macOS as my main OS, I can't really maintain it.

Should I let this PR be open, and wait for someone to come along (windows maintainer) or close it?

**Comment by skllcrn** [comment]
Thanks @albinekb! I suggest leaving it open for now and see if anyone wants to pick up where you left it.

**Comment by sindresorhus** [comment]
I'll close this. Can you open an issue instead here about adding Windows support?

**Comment by bdbch** [comment]
I use Windows as my main OS besides Arch Linux and could try to pick this up if you guys want me to. (And extend it further with Linux later)

**Comment by skllcrn** [comment]
At this point we won't be expanding the scope of Kap, however, if there was a high quality unofficial fork with support for another OS I'd argue that we should link to that. The reasoning being that maintaining this and any comparable open-source project is *a lot* of work, and the interest for most stops at the idea of doing it.

**Comment by levrik** [comment]
@skllcrn 
> with less people benefitting from it.
is purely wrong. I couldn't find a good free screen recording app for Windows. At the moment I use ShareX which works fine but it's a UI nightmare.
At work I use Kap on macOS. Love it and would really like to see it coming to Windows.

**Comment by skllcrn** [comment]
It's always great to hear from people that would like Kap on Windows. We'd love to see someone fork Kap and port it to Windows @levrik!

**Comment by shawwn** [comment]
+1 for Windows support @skllcrn @albinekb.  Gamedevs would thank you.

Right now your options for free recording on Windows are Windows Game Recorder (built in) or Streamlams / OBS. Both are less than ideal for recording promo videos and product walkthroughs.

**Comment by skllcrn** [comment]
Thanks for your comment @shawwn, but please refrain from posting "+1" and use 👍 on the related issues instead. When you do that we can sort issues based on reactions, making your voice heard and helping us prioritise. Remember that we work on a lot of projects, some of us in the tens and hundreds, and every single +1 is a notification/activity log item that just causes noise. 

> To show your interest your can leave a 👍 on #46 and wulkano/kap#456.

**Pull request closed without merging**